docs/guides/ssr.md

```diff
@@ -111,6 +180,95 @@ function Posts() {
 
 As demonstrated, it's fine to prefetch some queries and let others fetch on the queryClient. This means you can control what content server renders or not by adding or removing `prefetchQuery` for a specific query.
 
+#### With experimental `app` directory
+Wrap the `<QueryClientProvier>` and `<Hydrate>` components inside a Client Component that initializes the `QueryClient` your app will use on the client.
+
+- `QueryClient` should be created on an instance ref (or in React State ). **This ensures that data is not shared between different users and requests, while still only creating the QueryClient once per component lifecycle.**
+- The `dehydratedState` prop will contain your prefetched data along with `dataUpdatedAt` which provides the time the data was fetched on the server
+
+```tsx
+// app/react-query.jsx
+'use client'
+
+import { QueryClient, QueryClientProvider, Hydrate } from '@tanstack/react-query'
+
+export default function ReactQuery({dehydratedState, children}) {
+  const [queryClient] = React.useState(() => new QueryClient())
+  
+  return (
+    <QueryClientProvider client={queryClient}>
+      <Hydrate state={dehydratedState}>
```

Comment:
I think we should be able to decouple `QueryClientProvider` and `Hydrate`. The Provider needs to go at the top, once. Nothing changes here really.

`Hydrate` can appear multiple times in the app tree. It just takes data and puts it into the cache if it's newer, as opposed to `initialData`, which will only write data once to the cache if there is no data yet.

Reply:
Decoupling it is a really good point. In the current implementation, you might end up prefetching data at the top level for a leaf component that is conditionally rendered. Effectively, you could end up prefetching data that isn't needed in the end. Fetching data closer to the component that needs it could also allow you to parameterize that fetch with arguments that wouldn't be available at the top level.

I'll play around with these ideas a bit and see what kind of new implementation I can get working. Thinking something along the lines of a higher-order component that does the fetching and wraps its argument in a `Hydrate`.

Reply:
I'd say keep it simple for now, the API is probably going to change in the next RQ major anyway. If you wrap `QueryClientProvider` higher up in the tree, you can just do:

```jsx
'use client';
import { Hydrate } from '@tanstack/react-query'

export default function HydrateOnClient(dehydratedState) {
  return <Hydrate state={dehydratedState}>{children}</Hydrate>;
}
```

and use that directly in the Server Components. Don't think there is need for a HoC or something more complex here (at least yet).

Reply:
https://codesandbox.io/p/github/efilion/next13-react-query/decoupled-hydrate?file=%2Fapp%2Fpage.tsx

The HoC isn't strictly necessary, but it helps hide some of the complexity and keep things DRY.

To take advantage of the deduplication provided by QueryClient, the instance defined at the page level will need to be threaded down through to all components looking to prefetch data. That nullifies one of the advantages of the Hydration approach over the InitialData approach. It would be nice to have something akin to an IoC container that could provide a QueryClient as a request-scoped singleton, but I didn't see much in the way of options to accomplish that at the moment. InversifyJS and TSyringe don't fit the bill here. Likely a difficulty with the fact that being single-threaded, Node.js doesn't provide each request its own thread. Nest.js has a solution, so maybe Next.js could implement this itself down the line. Alternatively, can look at the Reader Monad pattern from functional programming to handle dependency injection.

Reply:
Yeah, that's a tricky one, clever solution!

I worry about putting that much complexity in the docs though, especially for an experimental approach, might fit better as an example to link to? There might be another way as well, not saying it's better but I thought I'd throw it out there.

So behind the scenes we have RSCs, client (server rendered) and client (browser rendered). For both of the client parts, it's important to have the same `queryClient` throughout the whole rendering which is why we want to hoist it to the top, but for the RSCs, your initial approach is actually fine where you create a new `queryClient` for each RSC (fine as in works, not as in pretty, DRY, fast etc). The reason for this is that since that data gets de/rehydrated anyway, we only need the `queryClient` as a way to be able to call `prefetchQuery` and get the right format back when we call `dehydrate`. The fetches themselves will be deduped behind the scenes anyway, so even if two RSCs call the same endpoint, it at least wont double fetch.

So for RSCs specifically, it's fine to duplicate the `queryClient` instead of passing it around. Ugly and not something we want long term, but might be easier and more concise to document, although possibly more confusing?

Reply:
> > but it helps hide some of the complexity and keep things DRY.
> 
> I guess another way of stating the above is that I'm not sure this is something we should aim to do at this point, I kind of like having the full extent of the experimental and raw nature of this fully visible, it becomes less of an "officially endorsed pattern" that way which is something that might bite us later (even with caveats, some devs will just copy-paste it). sweat_smile
> 
> Maybe I worry too much though.

I see what you mean. My HoC approach could be a better match for a blog post.

Reply:
https://codesandbox.io/p/github/efilion/next13-react-query/hydrated-content?file=%2Fapp%2FhydratedContent.tsx

Mainly looking to improve on line 12 inside `hydratedContent.tsx`. Not setting anything at all here might be the best approach for the moment. Might need to make a change to `dehydrate` to pick and choose which queries to dehydrate based on queryKey. Open to suggestions.